### ISSUE TITLE
Return the result of Q*K

### ISSUE THREAD

**Issue by CS123n** [facebookresearch/xformers#709]
# 🚀 Feature

Return the result of Q\*K.

## Motivation

Recently, numerous scholarly articles [1,2] have relied on the outcome of Q\*K to regulate subsequent generations. This approach has gained significant importance in research and practical applications. However, presently, they can only employ the original version to obtain the Q\*K result, which is slow.

[1] Prompt-to-Prompt Image Editing with Cross Attention Control
[2] Zero-shot Image-to-Image Translation

## Pitch

Return the result of Q\*K.

**Comment by danthe3rd** [comment]
Hi @CS123n - and thanks for the suggestion!
This is indeed interesting. However, from a technical perspective, what makes these implementations fast is that they don't need to store the Q*K result in memory (and memory bandwidth is quite slow).
We might lose most of the performance benefit if we were to store the result back in global memory.

**Comment by CS123n** [comment]
Hi, thanks a lot!
If storing the complete Q*K matrix proves to be a slow process, an alternative approach could be to calculate f(Q*K) instead. For instance, f could be defined as `torch.mean(Q*K, dim=1)` or `f=Q*K[:, j]`, which would fulfill the required purpose. Employing this method can effectively reduce the storage space needed while still preserving essential information about the matrix.
Is this possible?

**Comment by CS123n** [comment]
Wow, I see.
Replacing V with `torch.ones([1, 1, 1, ...])` or `torch.tensor([0, 0, 1, ...])` is indeed a good approach for `torch.mean` and `[:, j]`. Moreover, concatenating it with the original V can help in avoiding the need to compute Q*K twice!
Thanks a lot! @danthe3rd

**Comment by danthe3rd** [comment]
yes we support K and V having a different shape for the last dimension (you might need to pad V's last dimension to a multiple of 8)

**Comment by CS123n** [comment]
Get it! I will compare this with the original implementation.
